File: frontend/src/app/models/RolePermission.ts
import { BaseModel } from './BaseModel';

export interface RolePermissionDto extends BaseModel {
  systemRole: string;
  departmentId: string | null;
  departmentName?: string;
  systemModuleId: string;
  moduleName?: string;
  moduleKey?: string;

  canCreate: boolean;
  canRead: boolean;
  canUpdate: boolean;
  canDelete: boolean;
  canUpdateStatus: boolean;
}

export interface CreateRolePermissionRequest {
  systemRole: string;
  departmentId: string | null;
  systemModuleId: string;
  canCreate: boolean;
  canRead: boolean;
  canUpdate: boolean;
  canDelete: boolean;
  canUpdateStatus: boolean;
}

export interface UpdateRolePermissionRequest extends CreateRolePermissionRequest {
  id: string;
}

export interface UpsertRolePermissionDto {
  id?: string | null;
  systemRole: string;
  departmentId: string | null;
  systemModuleId: string;
  canCreate: boolean;
  canRead: boolean;
  canUpdate: boolean;
  canDelete: boolean;
  canUpdateStatus: boolean;
}

export interface ModuleRights {
  canCreate: boolean;
  canRead: boolean;
  canUpdate: boolean;
  canDelete: boolean;
  canUpdateStatus: boolean;
}

File: backend/Model/RolePermission.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace YLWorks.Model
{
    public class RolePermission : BaseEntity
    {
        [Key]
        public Guid Id { get; set; }

        [Required]
        [StringLength(50)]
        public string SystemRole { get; set; } = string.Empty; // e.g., "Management", "HOD", "Executive", "Support"

        public Guid? DepartmentId { get; set; }

        [ForeignKey("DepartmentId")]
        public Department? Department { get; set; }

        [Required]
        public Guid SystemModuleId { get; set; }

        [ForeignKey("SystemModuleId")]
        public SystemModule Module { get; set; } = null!;

        public bool CanCreate { get; set; } = false;
        public bool CanRead { get; set; } = false;   // (View)
        public bool CanUpdate { get; set; } = false; // (Edit)
        public bool CanDelete { get; set; } = false;
        public bool CanUpdateStatus { get; set; } = false; // (Approve/Reject Status changes)
    }

    public class UpsertRolePermissionDto
    {
        public Guid? Id { get; set; }
        [Required] public string SystemRole { get; set; } = string.Empty;
        public Guid? DepartmentId { get; set; }
        [Required] public Guid SystemModuleId { get; set; }
        public bool CanCreate { get; set; }
        public bool CanRead { get; set; }
        public bool CanUpdate { get; set; }
        public bool CanDelete { get; set; }
        public bool CanUpdateStatus { get; set; }
    }
}